queue/src/main/java/org/killbill/queue/DBBackedQueueWithInflightQueue.java

```diff
@@ -0,0 +1,327 @@
+
+        this.databaseTransactionNotificationApi = databaseTransactionNotificationApi;
+        databaseTransactionNotificationApi.registerForNotification(this);
+
+        // Metrics the size of the inflightQ
+        metricRegistry.register(MetricRegistry.name(DBBackedQueueWithInflightQueue.class, dbBackedQId, "inflightQ", "size"), new Gauge<Integer>() {
+            @Override
+            public Integer getValue() {
+                return inflightEvents.size();
+            }
+        });
+
+        this.transientInflightQRowIdCache = new TransientInflightQRowIdCache(queueId);
+    }
+
+    @Override
+    public void initialize() {
+
+        initializeInflightQueue();
+        log.info("{} Initialized with queueId={}, mode={}",
+                 DB_QUEUE_LOG_ID, queueId, config.getPersistentQueueMode());
+    }
+
+    @Override
+    public void close() {
+        databaseTransactionNotificationApi.unregisterForNotification(this);
+    }
+
+
+    @Override
+    public void insertEntryFromTransaction(final QueueSqlDao<T> transactional, final T entry) {
+        final Long lastInsertId = safeInsertEntry(transactional, entry);
+        if (lastInsertId == 0) {
+            log.warn("{} Failed to insert entry, lastInsertedId={}", DB_QUEUE_LOG_ID, lastInsertId);
+            return;
+        }
+
+        // The current thread is in the middle of  a transaction and this is the only times it knows about the recordId for the queue event;
+        // It keeps track of it as a per thread data. Very soon, when the transaction gets committed/rolled back it can then extract the info
+        // and insert the recordId into a blockingQ that is highly optimized to dispatch events.
+        transientInflightQRowIdCache.addRowId(lastInsertId);
+    }
+
+    private long pollEntriesFromInflightQ(final List<Long> result) {
+
+        long pollSleepTime = 0;
+        inflightEvents.drainTo(result, config.getMaxInFlightEntries());
+        if (result.isEmpty()) {
+            try {
+                long beforePollTime = System.nanoTime();
+                // We block until we see the first entry or reach the timeout (in which case we will rerun the doDispatchEvents() loop and come back here).
+                final Long entryId = inflightEvents.poll(INFLIGHT_POLLING_TIMEOUT_MSEC, TimeUnit.MILLISECONDS);
+                // Maybe there was at least one entry and we did not sleep at all, in which case this time is close to 0.
+                pollSleepTime = System.nanoTime() - beforePollTime;
+                if (entryId != null) {
+                    result.add(entryId);
+                }
+            } catch (final InterruptedException e) {
+                Thread.currentThread().interrupt();
+                log.warn("{} Got interrupted", DB_QUEUE_LOG_ID);
+                return 0;
+            }
+        }
+        return pollSleepTime;
+    }
+
+    @Override
+    public ReadyEntriesWithMetrics<T> getReadyEntries() {
```

Comment:
`ReadyEntriesWithMetrics` is only used for Bus with inflight queue. Is it worth the complexity computing it here? Couldn't the caller always compute the time spent (like in other cases)?

Reply:
No, impossible to compute from client because the blocking happens against the inflightQ deep in this class. This is different from notificationQ though -- previous remark about the `-1`.